// gatsby-remark-code-repls/default-redirect-template.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/inheritsLoose"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var GatsbyRemarkCodeReplsRedirect =
/*#__PURE__*/
function (_Component) {
  (0, _inheritsLoose2.default)(GatsbyRemarkCodeReplsRedirect, _Component);

  function GatsbyRemarkCodeReplsRedirect() {
    return _Component.apply(this, arguments) || this;
  }

  var _proto = GatsbyRemarkCodeReplsRedirect.prototype;

  _proto.componentDidMount = function componentDidMount() {
    this.form.submit();
  };

  _proto.render = function render() {
    var _this = this;

    var _this$props$pageConte = this.props.pageContext,
        action = _this$props$pageConte.action,
        payload = _this$props$pageConte.payload;
    return _react.default.createElement("form", {
      style: {
        paddingBottom: "50px"
      },
      ref: function ref(form) {
        _this.form = form;
      },
      action: action,
      method: "POST"
    }, _react.default.createElement("input", {
      type: "hidden",
      name: "data",
      value: payload
    }), _react.default.createElement("p", null, "Not automatically redirecting?"), _react.default.createElement("p", null, _react.default.createElement("input", {
      type: "submit",
      value: "Click here"
    })));
  };

  return GatsbyRemarkCodeReplsRedirect;
}(_react.Component);

GatsbyRemarkCodeReplsRedirect.propTypes = {
  pageContext: _propTypes.default.shape({
    action: _propTypes.default.string.isRequired,
    payload: _propTypes.default.object.isRequired
  }).isRequired
};
var _default = GatsbyRemarkCodeReplsRedirect;
exports.default = _default;
